lib/segment/src/index/query_optimization/rescore_formula/parsed_formula.rs

```diff
@@ -53,33 +57,118 @@ pub enum VariableId {
     Condition(ConditionId),
 }
 
-impl Expression {
-    pub fn new_div(left: Expression, right: Expression, by_zero_default: ScoreType) -> Self {
-        Expression::Div {
+impl ParsedExpression {
+    pub fn new_div(
+        left: ParsedExpression,
+        right: ParsedExpression,
+        by_zero_default: ScoreType,
+    ) -> Self {
+        ParsedExpression::Div {
             left: Box::new(left),
             right: Box::new(right),
             by_zero_default,
         }
     }
 
-    pub fn new_neg(expression: Expression) -> Self {
-        Expression::Neg(Box::new(expression))
+    pub fn new_neg(expression: ParsedExpression) -> Self {
+        ParsedExpression::Neg(Box::new(expression))
     }
 
     pub fn new_geo_distance(origin: GeoPoint, key: JsonPath) -> Self {
-        Expression::GeoDistance { origin, key }
+        ParsedExpression::GeoDistance { origin, key }
     }
 
-    #[cfg(feature = "testing")]
-    pub fn new_payload_id(path: &str) -> Self {
-        Expression::Variable(VariableId::Payload(JsonPath::new(path)))
+    pub fn new_payload_id(path: JsonPath) -> Self {
+        ParsedExpression::Variable(VariableId::Payload(path))
     }
 
     pub fn new_score_id(index: usize) -> Self {
-        Expression::Variable(VariableId::Score(index))
+        ParsedExpression::Variable(VariableId::Score(index))
     }
 
     pub fn new_condition_id(index: ConditionId) -> Self {
-        Expression::Variable(VariableId::Condition(index))
+        ParsedExpression::Variable(VariableId::Condition(index))
+    }
+}
+
+impl FromStr for VariableId {
+    type Err = String;
+
+    fn from_str(var_str: &str) -> Result<Self, Self::Err> {
+        let var_id = match var_str.strip_prefix("$") {
```

Comment:
Do we want to [trim](https://doc.rust-lang.org/std/primitive.str.html#method.trim) anywhere in this interface?

Reply:
I guess we could trim. Do we trim for JsonPath?